File: src/main/java/com/fendyk/clients/apis/MinecraftUserAPI.java
package com.fendyk.clients.apis;

import com.fendyk.API;
import com.fendyk.DTOs.LocationDTO;
import com.fendyk.DTOs.MinecraftUserDTO;
import com.fendyk.DTOs.SubscriptionRewardDTO;
import com.fendyk.DTOs.updates.UpdateMinecraftUserDTO;
import com.fendyk.clients.ClientAPI;
import com.fendyk.clients.fetch.FetchMinecraftUser;
import com.fendyk.clients.redis.RedisMinecraftUser;
import org.bukkit.Location;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.UUID;

public class MinecraftUserAPI extends ClientAPI<FetchMinecraftUser, RedisMinecraftUser, UUID, MinecraftUserDTO> {

    public MinecraftUserAPI(API api, FetchMinecraftUser fetch, RedisMinecraftUser redis) {
        super(api, fetch, redis);
    }

    /**
     * Gets the player's balance
     * @param player
     * @return
     */
    @Nullable
    public BigDecimal getPlayerBalance(UUID player) {
        MinecraftUserDTO minecraftUser = get(player);
        if(minecraftUser == null) return null;
        return BigDecimal.valueOf(minecraftUser.getQuanta());
    }

    /**
     * Gets the player from either redis or db
     * @param player
     * @return
     */
    @Nullable
    public MinecraftUserDTO get(UUID player) {
        MinecraftUserDTO dto = redis.get(player);
        cachedRecords.put(player, dto);
        return dto;
    }

    /**
     * Returns the cached player, pure for UI/Visuals that require loads of updates only.
     * @param player
     * @return
     */
    public MinecraftUserDTO getCached(Player player) {
        return getCached(player.getUniqueId());
    }

    /**
     * Updates the player on both redis and db
     * @param player
     * @return
     */
    public MinecraftUserDTO update(UUID player, UpdateMinecraftUserDTO minecraftUserDTO) {
        MinecraftUserDTO dto = fetch.update(player, minecraftUserDTO);
        cachedRecords.put(player, dto);
        return dto;
    }

    public boolean withDrawBalance(OfflinePlayer player, BigDecimal amount) {
        UUID uuid = player.getUniqueId();
        MinecraftUserDTO minecraftUser = get(uuid);
        if(minecraftUser == null) return false;

        BigDecimal oldAmount = BigDecimal.valueOf(minecraftUser.getQuanta());
        BigDecimal newAmount = oldAmount.subtract(amount).setScale(2, RoundingMode.HALF_EVEN);

        if(oldAmount.compareTo(amount) < 0) {
            return false;
        }

        UpdateMinecraftUserDTO update = new UpdateMinecraftUserDTO();
        update.setQuanta(newAmount.floatValue());
        return update(uuid, update) != null;
    }

    /**
     * Deposits money into account of the user
     * @param player
     * @param amount
     * @return
     */
    public boolean depositBalance(OfflinePlayer player, BigDecimal amount) {
        UUID uuid = player.getUniqueId();
        MinecraftUserDTO minecraftUser = get(uuid);
        if(minecraftUser == null) return false;

        BigDecimal oldAmount = BigDecimal.valueOf(minecraftUser.getQuanta());
        BigDecimal newAmount = oldAmount.add(amount).setScale(2, RoundingMode.HALF_EVEN);

        UpdateMinecraftUserDTO update = new UpdateMinecraftUserDTO();
        update.setQuanta(newAmount.floatValue());
        return update(uuid, update) != null;
    }

    /**
     * Updates the last location of the player.
     * @param player
     * @param location
     * @return
     */
    public boolean updateLastLocation(Player player, Location location) {
        UpdateMinecraftUserDTO update = new UpdateMinecraftUserDTO();
        update.setLastLocation(new LocationDTO(location));
        return update(player.getUniqueId(), update) != null;
    }

}

File: src/main/java/com/fendyk/clients/apis/MinecraftUserAPI.kt
package com.fendyk.clients.apis

import com.fendyk.DTOs.LocationDTO
import com.fendyk.DTOs.MinecraftUserDTO
import com.fendyk.DTOs.updates.UpdateMinecraftUserDTO
import com.fendyk.clients.ClientAPI
import com.fendyk.clients.fetch.FetchMinecraftUser
import com.fendyk.clients.redis.RedisMinecraftUser
import org.bukkit.Location
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player
import java.util.*
import java.util.concurrent.CompletableFuture

class MinecraftUserAPI(fetch: FetchMinecraftUser, redis: RedisMinecraftUser) :
        ClientAPI<FetchMinecraftUser, RedisMinecraftUser, UUID?, MinecraftUserDTO?>(fetch, redis) {
    /**
     * Gets the player's balance
     * @param player
     * @return
     */
    fun getPlayerBalance(player: OfflinePlayer): CompletableFuture<Double> {
        return CompletableFuture.supplyAsync {
            val awaitMinecraftUser = get(player.uniqueId)
            return@supplyAsync awaitMinecraftUser.join()?.quanta
        }
    }

    /**
     * Gets the player's balance
     * @param player
     * @return
     */
    fun hasEnoughBalance(player: OfflinePlayer, amount: Double): CompletableFuture<Boolean> {
        return CompletableFuture.supplyAsync {
            val minecraftUser = get(player.uniqueId).join()
            return@supplyAsync minecraftUser?.quanta != null && minecraftUser.quanta > amount
        }
    }

    /**
     * Gets the player from either redis or db
     * @param player
     * @return
     */
    fun get(player: UUID): CompletableFuture<MinecraftUserDTO?> {
        return CompletableFuture.supplyAsync {
            val minecraftUser = redis.get(player).join()
            cachedRecords[player] = minecraftUser
            return@supplyAsync minecraftUser
        }
    }

    /**
     * Returns the cached player, pure for UI/Visuals that require loads of updates only.
     * @param player
     * @return
     */
    fun getCached(player: Player): MinecraftUserDTO? {
        return getCached(player.uniqueId)
    }

    /**
     * Updates the player on both redis and db
     * @param player
     * @return
     */
    fun update(player: UUID, minecraftUserDTO: UpdateMinecraftUserDTO): CompletableFuture<MinecraftUserDTO?> {
        return CompletableFuture.supplyAsync {
            val minecraftUser: MinecraftUserDTO = fetch.update(player, minecraftUserDTO).join()
                    ?: throw Exception("Could not fetch your minecraft user")
            cachedRecords[player] = minecraftUser
            return@supplyAsync minecraftUser
        }
    }

    fun withDrawBalance(player: OfflinePlayer, amount: Double): CompletableFuture<Boolean> {
        return CompletableFuture.supplyAsync {
            val uuid: UUID = player.uniqueId
            val minecraftUser: MinecraftUserDTO? = get(uuid).join()

            val oldAmount: Double = minecraftUser?.quanta ?: 0.0
            val newAmount: Double = oldAmount - amount

            if (newAmount < 0) throw Exception("you cannot withdraw money that does not exist!")

            val updateDTO = UpdateMinecraftUserDTO()
            updateDTO.quanta = newAmount

            return@supplyAsync update(uuid, updateDTO).handleAsync<Boolean> { result: MinecraftUserDTO?, ex: Throwable? ->
                ex == null && result != null
            }.join()
        }
    }

    /**
     * Deposits money into account of the user
     * @param player
     * @param amount
     * @return
     */
    fun depositBalance(player: OfflinePlayer, amount: Double): CompletableFuture<Boolean> {
        return CompletableFuture.supplyAsync {
            val uuid: UUID = player.uniqueId
            val minecraftUser: MinecraftUserDTO? = get(uuid).join()

            val oldAmount: Double = minecraftUser?.quanta ?: 0.0
            val newAmount: Double = oldAmount + amount
            val updateDTO = UpdateMinecraftUserDTO()
            updateDTO.quanta = newAmount
            return@supplyAsync update(uuid, updateDTO).handleAsync<Boolean> { result: MinecraftUserDTO?, ex: Throwable? ->
                ex == null && result != null
            }.join()
        }
    }

    /**
     * Updates the last location of the player.
     * @param player
     * @param location
     * @return
     */
    fun updateLastLocation(player: Player, location: Location): CompletableFuture<Boolean> {
        return CompletableFuture.supplyAsync {
            val updateDTO = UpdateMinecraftUserDTO()
            updateDTO.lastLocation = LocationDTO(location)
            return@supplyAsync update(player.uniqueId, updateDTO).handleAsync<Boolean> { result: MinecraftUserDTO?, ex: Throwable? ->
                ex == null && result != null
            }.join()
        }
    }
}
